### PR TITLE
Avoid converting Decimal32/Decimal64 in `to_arrow` and `from_arrow` APIs

### PR DESCRIPTION
**Pull request by zeroshade**
## Description
Now that the Arrow format includes `Decimal32` and `Decimal64` data types, CUDF no longer needs to convert them to decimal128 when importing/exporting values via the `to_arrow` and `from_arrow` APIs. Instead we can just treat them like any other fixed-width data type and use the buffers directly.

This doesn't fully address https://github.com/rapidsai/cudf/issues/17080 as it doesn't make any changes to the Parquet side of things

This also incorporates the changes from https://github.com/rapidsai/cudf/pull/17405 which are needed for debug tests. That should get merged first, and then I can rebase this.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/interop/arrow_utilities.hpp [hunk 1]
```diff
@@ -86,5 +87,15 @@ std::unique_ptr<rmm::device_buffer> decimals_to_arrow(cudf::column_view input,
                                                       rmm::cuda_stream_view stream,
                                                       rmm::device_async_resource_ref mr);
 
+template <typename T, class Enable = void>
+struct DeviceType {
```
Comment (pmattione-nvidia) [inline cpp/src/interop/arrow_utilities.hpp:91]: There is already (e.g.) cudf::detail::DeviceType defined in copy_if.cuh. Do we really need this?

**Comment by vyasr** [comment]
Looks like a style fix is needed, and also you'll need to merge in the latest 25.02 to get all of CI passing again. I'll review in a few.

File: cpp/src/interop/arrow_utilities.hpp (hunk 1 above)
Comment (zeroshade) [reply cpp/src/interop/arrow_utilities.hpp:91]: oh nice! I didn't see that one there, we don't need this at all. I'll replace the usages of this with the one from `copy_if.cuh`

**Comment by zeroshade** [comment]
@vyasr Updated to fix the style issue and updated the branch

**Review by pmattione-nvidia — APPROVED** [review]

File: cpp/src/interop/to_arrow_schema.cpp [hunk 2]
```diff
@@ -102,7 +97,7 @@ int dispatch_to_arrow_type::operator()<numeric::decimal64>(column_view input,
                                                            ArrowSchema* out)
 {
   using DeviceType = int64_t;
-  return decimals_to_arrow<DeviceType>(input, out);
+  return decimals_to_arrow<DeviceType>(input, cudf::detail::max_precision<DeviceType>() - 1, out);
```
Comment (shrshi) [inline cpp/src/interop/to_arrow_schema.cpp:100]: Why are we subtracting 1 from the precision here?

**Review by shrshi — COMMENTED** [review]
Partial review - looks good so far! Still need to go through changes in `tests/interop`.
A few questions:

File: cpp/src/interop/arrow_utilities.hpp [hunk 3]
```diff
@@ -17,6 +17,7 @@
 #pragma once
 
 #include <cudf/types.hpp>
+#include <cudf/utilities/traits.hpp>
```
[the remark this answers is not in the document]
Comment (zeroshade) [reply cpp/src/interop/arrow_utilities.hpp:20]: no longer needed after the previous updates to use the existing `DeviceType` templates. I'll remove it.

File: cpp/src/interop/to_arrow_device.cu [hunk 4]
```diff
@@ -92,13 +92,15 @@ int set_buffer(std::unique_ptr<T> device_buf, int64_t i, ArrowArray* out)
...
+  template <typename T,
+            CUDF_ENABLE_IF(not is_rep_layout_compatible<T>() and not is_fixed_width<T>())>
   int operator()(cudf::column&&, rmm::cuda_stream_view, rmm::device_async_resource_ref, ArrowArray*)
   {
     CUDF_FAIL("Unsupported type for to_arrow_device", cudf::data_type_error);
   }
 
-  template <typename T, CUDF_ENABLE_IF(is_rep_layout_compatible<T>())>
+  // cover rep layout compatible and decimal types
+  template <typename T, CUDF_ENABLE_IF(is_rep_layout_compatible<T>() or is_fixed_width<T>())>
```
[the remark this answers is not in the document]
Comment (zeroshade) [reply cpp/src/interop/to_arrow_device.cu]: done

File: cpp/src/interop/to_arrow_schema.cpp (hunk 2 above)
Comment (zeroshade) [reply cpp/src/interop/to_arrow_schema.cpp:102]: Added a comment to explain the subtraction:

decimal64 max precision is 18 for Arrow, cudf uses `floor(bits * log(2) / log(10))` which comes out to 19, so we need to subtract 1.

File: cpp/tests/interop/from_arrow_device_test.cpp [hunk 5]
```diff
@@ -568,31 +573,37 @@ INSTANTIATE_TEST_CASE_P(FromArrowDeviceTest,
 template <typename T>
 using fp_wrapper = cudf::test::fixed_point_column_wrapper<T>;
 
-TEST_F(FromArrowDeviceTest, FixedPoint128Table)
+TYPED_TEST(FromArrowDeviceTestDecimalsTest, FixedPointTable)
 {
+  using T = TypeParam;
   using namespace numeric;
 
+  auto const precision = []() {
```
[the remark this answers is not in the document]
Comment (zeroshade) [reply cpp/tests/interop/from_arrow_device_test.cpp]: done

File: cpp/src/interop/arrow_utilities.hpp (hunk 1 above)
Comment (davidwendt) [reply cpp/src/interop/arrow_utilities.hpp]: This is being removed in https://github.com/rapidsai/cudf/pull/17520.
We may want to explore adding a new dispatch-type similar to [`dispatch_storage_type`](https://github.com/rapidsai/cudf/blob/13e983eafecad5a3d4053157febd714e40a410c3/cpp/include/cudf/utilities/type_dispatcher.hpp#L211) perhaps?

Comment (davidwendt) [reply cpp/src/interop/arrow_utilities.hpp]: Just noting that it does not include timestamp `rep` type. 
But that is not needed here so I think we should be good to use `dispatch_storage_type`

Comment (zeroshade) [reply cpp/src/interop/arrow_utilities.hpp]: updated

**Comment by zeroshade** [comment]
Hmm, it looks like the failing pyarrow tests might be a lack of support in pyarrow for decimal64/decimal32. I'll have to go fix that :frowning:

**Comment by zeroshade** [comment]
@davidwendt @vyasr With the release of Arrow v19, I've bumped the deps and this is finally ready. The python tests should pass now that pyarrow v19 added the decimal32 and decimal64 types properly, can you please take a look when you get a chance? 

Thanks!

File: cpp/tests/interop/from_arrow_host_test.cpp [hunk 6]
```diff
@@ -215,6 +221,244 @@ TYPED_TEST(FromArrowHostDeviceTestDurationsTest, DurationTable)
...
+{
+  using T = TypeParam;
+  using namespace numeric;
+
+  auto const precision = []() {
+    if constexpr (std::is_same_v<T, int64_t>)
+      return 18;
+    else
+      return cudf::detail::max_precision<T>();
+  }();
```
Comment (davidwendt) [inline cpp/tests/interop/from_arrow_host_test.cpp:237]: Use the `get_decimal_precision()` utility from `array_utils.hpp`?
https://github.com/rapidsai/cudf/pull/17422/files#diff-05280325f14b4019fc93c51fdd88ac57f5d58183aa008e619f853b117e0c2901R412

File: cpp/src/interop/from_arrow_host.cu [hunk 7]
```diff
@@ -20,6 +20,7 @@
 #include <cudf/column/column_view.hpp>
 #include <cudf/copying.hpp>
 #include <cudf/detail/copy.hpp>
+#include <cudf/detail/copy_if.cuh>
```
Comment (davidwendt) [inline cpp/src/interop/from_arrow_host.cu:23]:
```suggestion
```
I don't think this is used.

File: cpp/src/interop/to_arrow_host.cu [hunk 8]
```diff
@@ -18,6 +18,7 @@
 #include "decimal_conversion_utilities.cuh"
 
 #include <cudf/column/column_view.hpp>
+#include <cudf/detail/copy_if.cuh>
```
Comment (davidwendt) [inline cpp/src/interop/to_arrow_host.cu:21]:
```suggestion
```

File: cpp/tests/interop/from_arrow_host_test.cpp (hunk 6 above)
Comment (zeroshade) [reply cpp/tests/interop/from_arrow_host_test.cpp:237]: done

Comment (davidwendt) [reply cpp/tests/interop/from_arrow_host_test.cpp]: There are 4 more of these
https://github.com/rapidsai/cudf/pull/17422/files#diff-dba75f5fe9619a48cd45b04cbe3d568512a8df46cc228cb10e4ac465157a2ca8R458
https://github.com/rapidsai/cudf/pull/17422/files#diff-dba75f5fe9619a48cd45b04cbe3d568512a8df46cc228cb10e4ac465157a2ca8R491
https://github.com/rapidsai/cudf/pull/17422/files#diff-dba75f5fe9619a48cd45b04cbe3d568512a8df46cc228cb10e4ac465157a2ca8R527
https://github.com/rapidsai/cudf/pull/17422/files#diff-dba75f5fe9619a48cd45b04cbe3d568512a8df46cc228cb10e4ac465157a2ca8R563

Comment (zeroshade) [reply cpp/tests/interop/from_arrow_host_test.cpp]: bah, missed them cuz they were in another file. Fixed those spots too!

**Comment by davidwendt** [comment]
You'll need to update all the files in this PR to 2025. 
The style checker is not currently smart enough to detect only changed files within the commit.

**Comment by zeroshade** [comment]
Yup, i noticed the failing checks and ran pre-commit to update everything appropriately. should hopefully be good now! :)

**Comment by zeroshade** [comment]
ugh, don't know what happened there. but *now* it's correct

**Comment by zeroshade** [comment]
@davidwendt updated and fixed the failing unit tests

**Comment by zeroshade** [comment]
Okay, looks like the python wheel failures were due to using decimal32/decimal64 types directly when testing with pyarrow < 19. This is because now that we aren't converting decimal32/decimal64 to decimal128, those schema format strings are no longer compatible with older versions of pyarrow (as the decimal32/decimal64 data types were only added in pyarrow v19.0.0) so the format string "d:18,3,32" is not valid in older version of pyarrow. 

So I've added logic to skip those tests when testing with older versions of pyarrow (it still tests the decimal128 case). That should fix the failing tests.

**Comment by davidwendt** [comment]
How about the Java test errors? Looks like it may be a similar problem.

**Comment by zeroshade** [comment]
I completely missed that and thought that the Java failures were unrelated. I'll see if i can figure out a solution to those, but I will admit that Java is not my strong suit haha

**Comment by zeroshade** [comment]
@davidwendt looks like arrow-java doesn't support Decimal32/Decimal64 yet, so I've disabled the failing tests. they can be re-enabled when arrow-java adds support for those types.

**Comment by zeroshade** [comment]
Looking at the logs for the failing python wheel test, I don't think it's related to or caused by this change. At least I don't know how it would be related given the errors and that the tests for cuda 12.5 succeed.

If it is caused or related to this change, I think i'll need help diagnosing the issue there....

**Comment by davidwendt** [comment]
> Looking at the logs for the failing python wheel test, I don't think it's related to or caused by this change. At least I don't know how it would be related given the errors and that the tests for cuda 12.5 succeed.
> 
> If it is caused or related to this change, I think i'll need help diagnosing the issue there....

Ok. I'll ask some of the python experts to take a look.

**Comment by davidwendt** [comment]
I thought the error was related to recent numba change but that appears to be not the case.
The failing test is here: https://github.com/rapidsai/cudf/blob/1f0f51f96b79edd820e81343ca521c684b1f4918/python/cudf/cudf/tests/test_parquet.py#L4410
It certainly looks decimal related I think. You can recreate the error(s) running the following in the `cudf/python/cudf/cudf/tests` directory: `python -m pytest test_parquet.py -k'test_parquet_bloom_filters'`
Perhaps @mhaseeb123  can help with this as well?

**Comment by davidwendt** [comment]
It fails under the CI GHA 11.8 with ARM container setup. I was able recreate it locally using steps from here: https://docs.rapids.ai/resources/reproducing-ci/ but was not able to get further.

File: dependencies.yaml [hunk 9]
```diff
@@ -474,16 +474,16 @@ dependencies:
     common:
       - output_types: [conda]
         packages:
-          - pyarrow>=14.0.0,<19.0.0a0
+          - pyarrow>=14.0.0,<=19.0.0
```
Comment (bdice) [inline dependencies.yaml:477]: Let's use the next major version as an upper bound, to permit `19.x.y` releases.
```suggestion
          - pyarrow>=14.0.0,<20.0.0a0
```

File: dependencies.yaml [hunk 10]
```diff
@@ -474,16 +474,16 @@ dependencies:
...
         packages:
           # pyarrow 17.0.0 wheels have a subtle issue around threading that
           # can cause segmentation faults around imports on arm. It appears to
           # be highly dependent on the exact build configuration, so we'll just
           # avoid 17.0.0 for now unless we observe similar issues in future
           # releases as well.
-          - pyarrow>=14.0.0,<19.0.0a0; platform_machine=='x86_64'
-          - pyarrow>=14.0.0,<19.0.0a0,!=17.0.0; platform_machine=='aarch64'
+          - pyarrow>=14.0.0,<=19.0.0; platform_machine=='x86_64'
+          - pyarrow>=14.0.0,<=19.0.0,!=17.0.0; platform_machine=='aarch64'
```
Comment (bdice) [inline dependencies.yaml:486]:
```suggestion
          - pyarrow>=14.0.0,<20.0.0a0; platform_machine=='x86_64'
          - pyarrow>=14.0.0,<20.0.0a0,!=17.0.0; platform_machine=='aarch64'
```

**Review by bdice — CHANGES_REQUESTED** [review]

**Comment by bdice** [comment]
@zeroshade @davidwendt Let's see if that error is fixed by https://github.com/rapidsai/cudf/pull/17758. We know there are some oddities that were occurring because storage was not correctly aligned for the bloom filters. Perhaps this PR makes those issues manifest in a different way than we saw earlier.

File: dependencies.yaml (hunk 9 above)
Comment (zeroshade) [reply dependencies.yaml:477]: good point!

**Comment by zeroshade** [comment]
ugh, sorry. i missed a file when updating after running pre-commit. this time I actually *did* push the fixes for the style checks. 

It would be great if it still ran the unit tests and build even if the style check failed so that we don't have to keep restarting to determine if the other issues were addressed

**Comment by zeroshade** [comment]
@mhaseeb123 @davidwendt looks like the same Arm-based bloom filter issue, I still don't know how it could be related to the decimal changes here though :frowning:

**Comment by mhaseeb123** [comment]
>@mhaseeb123 @davidwendt looks like the same Arm-based bloom filter issue, I still don't know how it could be related to the decimal changes here though 😦

I investigated this. TLDR; It's not related to bloom filters but most likely something else. Here are some of my findings. 
1. I checked out the CI artifacts with help from @davidwendt.
2. As David mentioned, the failure point is here: https://github.com/rapidsai/cudf/blob/1f0f51f96b79edd820e81343ca521c684b1f4918/python/cudf/cudf/tests/test_parquet.py#L4410
3. At this line, we are reading [this](https://github.com/rapidsai/cudf/blob/branch-25.02/python/cudf/cudf/tests/data/parquet/mixed_card_ndv_100_chunk_stats.snappy.parquet) parquet file which only has stats and no bloom filters.
4. At this point, I thought something went wrong with the stats based predicate pushdown but I also noticed that the following fails (from the same docker container from the CI):
```python
import pyarrow.parquet as pq
pq.read_table("/repo/python/cudf/cudf/tests/data/parquet/mixed_card_ndv_100_chunk_stats.snappy.parquet")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/pyenv/versions/3.12.8/lib/python3.12/site-packages/pyarrow/parquet/core.py", line 1843, in read_table
    return dataset.read(columns=columns, use_threads=use_threads,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/pyenv/versions/3.12.8/lib/python3.12/site-packages/pyarrow/parquet/core.py", line 1485, in read
    table = self._dataset.to_table(
            ^^^^^^^^^^^^^^^^^^^^^^^
  File "pyarrow/_dataset.pyx", line 574, in pyarrow._dataset.Dataset.to_table
  File "pyarrow/_dataset.pyx", line 3865, in pyarrow._dataset.Scanner.to_table
  File "pyarrow/error.pxi", line 155, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 92, in pyarrow.lib.check_status
OSError: Repetition level histogram size mismatch
```
and we actually get an abort if we try to add a filter to it:

```python
import pyarrow.parquet as pq
pq.read_table("/repo/python/cudf/cudf/tests/data/parquet/mixed_card_ndv_100_chunk_stats.snappy.parquet", filters=[("str", "=", "FINDME")])
terminate called after throwing an instance of 'parquet::ParquetException'
  what():  Repetition level histogram size mismatch
Aborted (core dumped)
```
I tried finding some clues about this and came across https://github.com/apache/arrow/issues/45283. Looks like both issues are related since the file: `mixed_card_ndv_100_chunk_stats.snappy.parquet` has also been generated using arrow-rs
5. Anyhow back to the original investigation, then I changed the parameters to the affected pytest https://github.com/rapidsai/cudf/blob/1f0f51f96b79edd820e81343ca521c684b1f4918/python/cudf/cudf/tests/test_parquet.py#L4392 
to simply the following to ensure that predicates involving decimals were not the root cause but still got the same failure that we see in the CI
```
@pytest.mark.parametrize(
    "predicate,expected_len",
    [
        ([("str", "==", "FINDME")], 2),
    }
)
```
6. Further, if we remove the `filters=predicate` argument to: https://github.com/rapidsai/cudf/blob/1f0f51f96b79edd820e81343ca521c684b1f4918/python/cudf/cudf/tests/test_parquet.py#L4410 the test starts running and correctly asserts at `assert_eq()` as expected.

To summarize; the abort we are seeing is likely originating from arrow. How we are getting to that point is what I am not fully sure yet is where and how exactly we are interacting with arrow during (likely) `stats` based row group filtering. We should also check if the fix for https://github.com/apache/arrow/issues/45283 has been pulled into libcudf to narrow down the fix to this.

**Comment by zeroshade** [comment]
@mhaseeb123 @davidwendt If the identified issue is the cause of the CI failure, it's odd that it would only occur on the ARM image and not on the AMD64 one. That said, it seems like you were able to confirm that the issue is definitely not related to the decimal changes at all, right? 

I'm going to try to reproduce this failure locally via the suggested image and then see if merely using a version of arrow-cpp with that fix is sufficient to fix it. In the meantime, I'll see if I can influence us getting a v19.0.1 patch release of Arrow/Pyarrow with the fix in general.

**Comment by zeroshade** [comment]
@mhaseeb123 @davidwendt Okay, I was able to explicitly confirm that if I install a version of pyarrow that contains the fix (easiest route is to install the nightly package as described in the [docs](https://arrow.apache.org/docs/developers/python.html#installing-nightly-packages)) the issue goes away. It isn't related at all to the version of arrow-cpp that libcudf has, as the tests install pyarrow which has arrow-cpp bundled for itself. 

So this proves and confirms the CI failure is entirely related to the issue you linked and not caused by any of the decimal related code here. Essentially, it's because this bumps pyarrow to v19 which has the bug you identified.

I guess we need to wait until a patch is released with the fix before we can merge this then? :frowning:

**Comment by davidwendt** [comment]
There is this PR https://github.com/rapidsai/cudf/pull/17794
Could we have this include the fix somehow?
I'm ok disabling or working around the test until the patch is ready if you are Haseeb.

**Comment by mhaseeb123** [comment]
> Could we have this include the fix somehow?
Not sure, I think we would need a patch release version but I may be wrong here.

> I'm ok disabling or working around the test until the patch is ready if you are Haseeb.
I am okay with the work around I proposed above because that way we can at least keep the bloom filter part of the test.

**Comment by davidwendt** [comment]
@zeroshade Can you try out Haseeb's suggestion from https://github.com/rapidsai/cudf/pull/17422#issuecomment-2611212357 to see if this gets around the build issue for this PR?

**Comment by mhaseeb123** [comment]
> @zeroshade Can you try out Haseeb's suggestion from [#17422 (comment)](https://github.com/rapidsai/cudf/pull/17422#issuecomment-2611212357) to see if this gets around the build issue for this PR?

Yes, that would be super helpful to double-check. I also just locally verified on my end and it was good 🙂

**Comment by bdice** [comment]
It sounds like 19.0.0 is required for this PR, but 19.0.0 introduces a bug? I am okay skipping the test temporarily if cuDF behaves correctly but the reference implementation in PyArrow that we compare against is incorrect. I don’t want to accept 19.0.0 without a patch if it introduces incorrect cuDF behavior.

**Comment by galipremsagar** [comment]
> It sounds like 19.0.0 is required for this PR, but 19.0.0 introduces a bug? I am okay skipping the test temporarily if cuDF behaves correctly but the reference implementation in PyArrow that we compare against is incorrect. I don’t want to accept 19.0.0 without a patch if it introduces incorrect cuDF behavior.

I updated https://github.com/rapidsai/cudf/pull/17794 to skip the failing test only on arm, because that is what seems to be happening.

**Comment by zeroshade** [comment]
Thanks @galipremsagar I've added that to this change here.

@mhaseeb123 @davidwendt I've skipped the test here, so we should be good.

> It sounds like 19.0.0 is required for this PR, but 19.0.0 introduces a bug? I am okay skipping the test temporarily if cuDF behaves correctly but the reference implementation in PyArrow that we compare against is incorrect. I don’t want to accept 19.0.0 without a patch if it introduces incorrect cuDF behavior.

@bdice that's exactly the case, 19.0.0 is required for this PR but introduces a bug in the reference implementation that you're comparing against. It is not introducing any incorrect cudf behavior.

**Comment by zeroshade** [comment]
Ooops forgot to push the change. I've pushed the updates now, and should be good to run the CI (hopefully)

**Comment by bdice** [comment]
@zeroshade Let's merge #17794 first, it is nearly done -- just waiting on CI to complete. I asked for a slight modification to the skip criteria there, so there will be minor merge conflicts with this PR.

File: conda/recipes/cudf/meta.yaml [hunk 11]
```diff
@@ -81,7 +81,7 @@ requirements:
     - numba-cuda >=0.2.0,<0.3.0a0
     - numba >=0.59.1,<0.61.0a0
     - numpy >=1.23,<3.0a0
-    - pyarrow>=14.0.0,<18.0.0a0
+    - pyarrow>=14.0.0,<=19.0.0
```
Comment (galipremsagar) [inline conda/recipes/cudf/meta.yaml:84]:
```suggestion
    - pyarrow>=14.0.0,<20.0.0a0
```

File: conda/recipes/pylibcudf/meta.yaml [hunk 12]
```diff
@@ -77,7 +77,7 @@ requirements:
     - typing_extensions >=4.0.0
     - pandas >=2.0,<2.2.4dev0
     - numpy >=1.23,<3.0a0
-    - pyarrow>=14.0.0,<18.0.0a0
+    - pyarrow>=14.0.0,<=19.0.0
```
Comment (galipremsagar) [inline conda/recipes/pylibcudf/meta.yaml:80]:
```suggestion
    - pyarrow>=14.0.0,<20.0.0a0
```

File: python/cudf/cudf/tests/test_scalar.py [hunk 13]
```diff
@@ -381,6 +397,50 @@ def test_scalar_invalid_implicit_conversion(cls, dtype):
...
+def test_device_scalar_direct_construction(value, decimal_type):
+    value = cudf.utils.dtypes.to_cudf_compatible_scalar(value)
+
+    dtype = (
+        value.dtype
+        if not isinstance(value, Decimal)
+        else decimal_type._from_decimal(value)
+    )
+
+    s = cudf._lib.scalar.DeviceScalar(value, dtype)
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_scalar.py:430]:
```suggestion
    s = cudf.Scalar(value, dtype)
```

File: cpp/cmake/thirdparty/patches/nanoarrow_override.json [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+{
+  "packages" : {
+    "nanoarrow" : {
+      "version" : "0.7.0.dev",
+      "git_url" : "https://github.com/apache/arrow-nanoarrow.git",
+      "git_tag" : "4bf5a9322626e95e3717e43de7616c0a256179eb",
+      "git_shallow" : false,
+      "patches" : [
+        {
+          "file" : "${current_json_dir}/nanoarrow_clang_tidy_compliance.diff",
```
Comment (bdice) [inline cpp/cmake/thirdparty/patches/nanoarrow_override.json:11]: Is this patch still relevant? I don’t see it here. Can this override be removed?

**Comment by zeroshade** [comment]
@galipremsagar @davidwendt @bdice I don't see where the `py-polars` tests that are failing are in the repo, given that the failure us a `pyarrow_slice_roundtrip` i'm guessing it's the same issue as the others in that it's hitting the bug that needs to get patched? 

Btw: we are working on pushing out a 19.0.1 patch fix for pyarrow to fix the issue, but it'll be a few days before that happens still.

**Comment by davidwendt** [comment]
Yes, sorry Matt. We are having some build issues in CI that we are trying to work through that are showing up here as well.
Hopefully we'll be able to CI working again soon.

**Comment by bdice** [comment]
I think CI issues should be cleared up now. I triggered a new run.

File: cpp/cmake/thirdparty/patches/nanoarrow_override.json (hunk 14 above)
Comment (bdice) [reply cpp/cmake/thirdparty/patches/nanoarrow_override.json:11]: @zeroshade Can we remove this? I'm okay with merging this PR and opening a follow-up for cleanup, since CI has passed.

Comment (bdice) [reply cpp/cmake/thirdparty/patches/nanoarrow_override.json:11]: Looks like #17618 removed this patch. I'll merge this PR and file a follow-up to delete this override.

**Review by bdice — APPROVED** [review]

**Comment by bdice** [comment]
Requesting reviews from the @rapidsai/cudf-java-codeowners and @rapidsai/cudf-python-codeowners teams.

File: cpp/cmake/thirdparty/patches/nanoarrow_override.json (hunk 14 above)
Comment (bdice) [reply cpp/cmake/thirdparty/patches/nanoarrow_override.json:11]: I realized we are missing some reviews. If we need to push any changes to this branch for other reviews, we should go ahead and delete this. Otherwise I think merging this and removing the overrides in a follow-up is the best case scenario.

**Review by revans2 — COMMENTED** [review]
I need to run some tests on Spark to know if this is a breaking change or not. Spark uses java-arrow (a version that we do not control) for interchange with python. I am concerned that the version of python-arrow that is used is also not compatible with this change, or if it is, then it will send back decimal values that are not compatible with the java version.

File: python/cudf/cudf/tests/test_reductions.py [hunk 15]
```diff
@@ -204,10 +301,34 @@ def test_max(dtype, nelem):
...
+                reason="decimal64 format string only supported in pyarrow >=19",
+            ),
+        ),
+        pytest.param(
+            Decimal32Dtype(6, 3),
+            marks=pytest.mark.skipif(
+                pa._generated_version.version_tuple[0] < 19,
+                reason="decimal32 format string only supported in pyarrow >=19",
+            ),
+        ),
```
Comment (wence-) [inline python/cudf/cudf/tests/test_reductions.py:331]: Question: does this mean that decimal32/64 are no longer supported unless the user has pyarrow >= 19? Or do I misunderstand what is going on?

**Review by wence- — APPROVED** [review]
Approving python changes with one question about what the skipped tests mean for decimal32/64 support when the user doesn't have pyarrow 19.

**Review by galipremsagar — APPROVED** [review]

File: python/cudf/cudf/tests/test_reductions.py (hunk 15 above)
Comment (zeroshade) [reply python/cudf/cudf/tests/test_reductions.py:331]: Nope, you're correct. (Well, they are still supported, but you can't convert to pyarrow unless they have >= 19)

The only way to avoid that would be to add a flag at the C++ level to tell it to fall back and upcast the decimal32/64 back to decimal128 and then conditionally set that flag if they are using pyarrow <19.

**Review by revans2 — APPROVED** [review]
I am approving this for the java side of things, but I am not happy with it, because it paints us into a corner. We support Spark 3.2.0+ right now. The python interface for Spark 3.2.0 through 3.4.x requires at least pyarrow 1.0.0. Spark 3.5.0+ currently requires at least pyarrow 4.0.0.  

Neither of these are are compatible with [pyarrow >= 19](https://github.com/rapidsai/cudf/pull/17422#discussion_r1932418403). So it will not work for us to support any kind of decimal transfer. The reason I am approving this because the Spark RAPIDS plugin doesn't currently support decimal for python UDFs. It looks like that was a missed feature when we added in decimal support. I made a 3 line change to our plugin to add in decimal support, tested it, and it worked just fine without this patch. With this patch it fails.

I would love the ability to serialize the data in a way that is compatible with older versions of arrow, but it is not a requirement right now.

File: python/cudf/cudf/tests/test_reductions.py (hunk 15 above)
Comment (bdice) [reply python/cudf/cudf/tests/test_reductions.py:331]: Does this tie in with the Java discussion here? https://github.com/rapidsai/cudf/pull/17422#pullrequestreview-2579257680

Do we need a way to make this support older pyarrow versions with the previous behavior?

**Comment by zeroshade** [comment]
@revans2 The only way for it to be compatible with older versions of arrow would be to have it perform the cast/copy/conversion to decimal128. 

That said, what makes those incompatible with pyarrow >= 19 if they only specify the minimum being 1.0 / 4.0? Do they have a maximum version they support? If the spark interfaces don't work with the newer versions of pyarrow, that's a problem and should be addressed by either us in the Arrow project or by Spark.

**Comment by revans2** [comment]
> That said, what makes those incompatible with pyarrow >= 19 

We don't technically have any control over what the customer's environment is setup as beyond the requirements that we have the cuda drivers installed and our plugin is on the classpath. Even the cuda runtime can be too much for some of them (I'm looking at you dataproc). It is highly unlikely that anyone is running with the 5 year old pyarrow 1.0.0, but it is possible (Spark 3.2.0, which we have customers running is 4 years old). We want our plugin to be compatible with whatever setup that they have already setup and tested.

It is, however, very likely that our customers are not running on the pyarrow 19.0.0, which is 13 days old as of writing this. The environments we run in are often set by CSPs and large enterprises where they find something that works and don't touch it for years except when the security people force them to. I want to reduce the friction for adopting our product as much as possible.

But you do have a point. I can work around it myself and cast the decimal values to DECIMAL_128 before I send them. It would slow things down, especially for nested data, but it would work.

**Review by davidwendt — APPROVED** [review]

File: python/cudf/cudf/tests/test_reductions.py (hunk 15 above)
Comment (vyasr) [reply python/cudf/cudf/tests/test_reductions.py:331]: Where exactly is it failing? I would expect that the computation of `got` proceeds fine, but `assert_eq` fails because we do the comparison by converting the cudf object to host via `to_arrow`. Is that right?

If so, I am OK with skipping this. I think the same argument that @zeroshade made in https://github.com/rapidsai/cudf/pull/17422#issuecomment-2619919872 applies here too: users can still do this, but they have to perform the cast themselves first. I don't view that as a negative since it makes explicit a copy that was previously happening implicitly anyway. We could make the test "work" by casting manually, but at that point we'd have to also inject extra information about the source dtype pre-cast since that information would be lost.

Relatedly, I expect that in the next release we'll move to pylibcudf exclusively exposing the capsule interface rather than pyarrow types in its interop, at which point we'll be even less tied to a particular pyarrow version's specific support for the Arrow spec than we are now.

Comment (vyasr) [reply python/cudf/cudf/tests/test_reductions.py:331]: If we wanted to be extra helpful to users we could insert a dtype check in `to_arrow` and raise a more informative error. Given the upcoming changes I'm inclined to say that's not worthwhile, though.

**Review by vyasr — APPROVED** [review]
There's one thread that @wence- asked about decimal32/64 support in the Python test suite that we should make sure we come to a resolution on that everyone is happy with, but aside from that this looks great. Thanks as always for the attention to detail in the test suite!

File: python/cudf/cudf/tests/test_reductions.py (hunk 15 above)
Comment (zeroshade) [reply python/cudf/cudf/tests/test_reductions.py:331]: @vyasr it fails on the pyarrow side because it doesn't recognize the schema format string coming from the C ABI. Older versions of pyarrow reject the bit width of 32 and 64.

> Relatedly, I expect that in the next release we'll move to pylibcudf exclusively exposing the capsule interface rather than pyarrow types in its interop, at which point we'll be even less tied to a particular pyarrow version's specific support for the Arrow spec than we are now.

Has this work begun yet? I also really want to be exposing the` to_arrow_device` and `from_arrow_device` in Python. Currently, pylibcudf forces copy between CPU and GPU for arrow.

Comment (vyasr) [reply python/cudf/cudf/tests/test_reductions.py:331]: Right and I'm assuming that pyarrow interaction only happens in the `assert_eq` call, so I don't think there's anything that we need to change here.

I started working on #16104 because I want the lifetime management handled properly before we update pylibcudf. That's high on my priority list for 25.04. Once that is done, I will update #15370 accordingly. If you want to help me out with #16104 I'm happy to push something up sooner rather than later for you to play with!

**Comment by vyasr** [comment]
Thanks everyone for the reviews here! I think we're all set now.
